src/mame/sinclair/scorpion.cpp

```diff
@@ -558,16 +651,242 @@ void scorpiontb_state::scorpion_ioext(address_map &map)
+ * Scorpion GMX
+ * ***************************************************************************/
+void scorpiongmx_state::scorpion_update_memory()
+{
+	scorpiontb_state::scorpion_update_memory();
+	if (BIT(m_port_1ffd_data, 2))
+	{
+		m_bank_rom[0]->set_entry((m_prof_plane << 2) | ROM_PAGE_DOS);
+		m_bank0_rom.disable();
+		m_beta->enable();
+		LOGMEM("... ROM0(%x)", m_bank_rom[0]->entry());
+	}
+	m_bank_ram[2]->set_entry((m_port_78fd_data ^ 2) % m_ram_banks);
+	LOGMEM("... RAM2(%x)", m_bank_ram[2]->entry());
+	m_bank_ram[3]->set_entry((((m_port_dffd_data & 7) << 4) | ((m_port_1ffd_data & 0x10) >> 1) | (m_port_7ffd_data & 0x07)) % m_ram_banks);
+	LOGMEM(", RAM3(%x)\n", m_bank_ram[3]->entry());
+}
+
+void scorpiongmx_state::spectrum_update_screen(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
+{
+	if (!m_gfx_ext)
+		return scorpiontb_state::spectrum_update_screen(screen, bitmap, cliprect);
+
+	// 640x200 ...
+	spectrum_update_border(screen, bitmap, cliprect);
+
+	const bool invert_attrs = u64(screen.frame_number() / m_frame_invert_count) & 1;
+	const u8 *screen_location = m_ram->pointer() + ((BIT(m_port_7ffd_data, 3) ? 0x3b : 0x39) << 14);
+	for (u16 vpos = cliprect.top(); vpos <= cliprect.bottom(); vpos++)
+	{
+		const u16 y = (vpos - get_screen_area().top() + (((m_scroll_hi << 8) | m_scroll_lo) / 80)) % (25 * 8);
+		for (u16 hpos = cliprect.left() & 0xfff8; hpos <= cliprect.right();)
+		{
+			const u16 x = hpos - get_screen_area().left();
+			const u8 *scr = screen_location + (y * 80) + (x >> 3);
+
+			const u8 attr = *(scr + (0x40 << 14));
+			const u16 fg = ((attr >> 3) & 0x08) | (attr & 0x07);
+			const u16 bg = (attr >> 3) & 0x0f;
+			const bool invert = (attr & 0x80) && invert_attrs;
+
+			const u8 chunk = *scr;
+			for (u8 i = 0x80; i; i >>= 1)
+				bitmap.pix(vpos, hpos++) = (chunk & i)
+					? (invert ? bg : fg)
+					: (invert ? fg : bg);
+		}
+	}
+}
+
+rectangle scorpiongmx_state::get_screen_area()
+{
+	return m_gfx_ext
+		? rectangle { 80, 80 + 639, 80, 80 + 199 }
+		: scorpiontb_state::get_screen_area();
+}
+
+void scorpiongmx_state::do_nmi()
+{
+	if (m_magic_disabled)
+	{
+		LOGIO("NMI ignored as Magic disabled");
+		m_nmi_pending = 0;
+	}
+	else
+	{
+		m_port_7afd_lock_data = port_7afd_r() & 0x7f;
+		m_port_7efd_lock_data = port_7efd_r() & 0x4f;
+		LOGIO("NMI: registers locked 7afd=%02x, 7efd=%02x\n", m_port_7afd_lock_data, m_port_7efd_lock_data);
+		scorpion_state::do_nmi();
+		m_magic_lock = 1;
+	}
+}
+
+void scorpiongmx_state::global_cfg_w(u8 data)
+{
+	m_port_00_data = data;
+	if (BIT(data, 5)) // BLKEXT
+	{
+		LOGIO("GMX ports disabled\n");
+		m_io_gmx.disable();
+	}
+	else
+		m_io_gmx.select(0);
+
+	if (BIT(data, 3))
+	{
+		m_magic_shift = 0x88 | (data & 7);
+		if (!GMX_FIXROM)
+			m_maincpu->reset();
+	}
+}
+
+u8 scorpiongmx_state::port_78fd_r()
+{
+	u8 tmp = (BIT(m_port_fe_data, 1) << 7) // BRD1
+		| (m_port_78fd_data & 0x7f);
+	tmp |= (m_magic_shift & 1);
+	m_magic_shift >>= 1;
+	return tmp;
+}
```

Comment:
You need to check `!machine().side_effects_disabled()` before updating `m_magic_shift` or bad things will happen.